agent/lib/kontena/websocket_client.rb

```diff
@@ -61,57 +46,94 @@ def connecting?
       @connecting
     end
 
+    def rpc_server
+      Celluloid::Actor[:rpc_server]
+    end
+    def rpc_client
+      Celluloid::Actor[:rpc_client]
+    end
+
+    def start
+      every(1.0) do
+        connect if !connected? unless connecting?
+      end
+
+      every(KEEPALIVE_INTERVAL) do
+        keepalive if connected?
+      end
+    end
+
     def connect
-      return if connecting?
-      @connected = false
-      @connecting = true
       info "connecting to master at #{api_uri}"
       headers = {
-          'Kontena-Grid-Token' => self.api_token.to_s,
+          'Kontena-Grid-Token' => @api_token.to_s,
           'Kontena-Node-Id' => host_id.to_s,
           'Kontena-Version' => Kontena::Agent::VERSION,
           'Kontena-Node-Labels' => labels,
           'Kontena-Connected-At' => Time.now.utc.strftime(STRFTIME),
       }
-      @ws = Faye::WebSocket::Client.new(self.api_uri, nil, {headers: headers})
 
-      notify_actors('websocket:connect', self)
+      @ws = Kontena::Websocket::Client.new(@api_uri,
+        headers: headers,
+        ssl_verify: true,
+      )
 
-      @ws.on :open do |event|
-        on_open(event)
-      end
-      @ws.on :message do |event|
-        on_message(event)
-      end
-      @ws.on :close do |event|
-        on_close(event)
-      end
-      @ws.on :error do |event|
-        on_error(event)
-      end
+      @connecting = true
+
+      # service the websocket reads, and emit events
+      async.connect_ws @ws
+
+
+      publish('websocket:connect', Actor.current)
+    end
+
+    # @param ws [Kontena::Websocket::Client]
+    def connect_ws(ws)
+      actor = Actor.current
+
+      # run the blocking websocket client read in a separate thread
+      # raises on errors, crashing the actor
+      defer {
+        ws.connect! do |ws|
+          # These are called from the read_ws -> defer thread, proxy back to actor
+          ws.on :open do |event|
+            actor.on_open(event)
+          end
+          ws.on :message do |event|
+            actor.on_message(event)
+          end
+          ws.on :close do |event|
+            actor.on_close(event)
+          end
+          ws.on :error do |event|
+            actor.on_error(event)
+          end
+        end
+
+        ws.read
+      }
     end
 
     ##
     # @param [String, Array] msg
     def send_message(msg)
-      EM.next_tick {
-        begin
-          @ws.send(msg) if @ws
-        rescue
-          error "failed to send message"
-        end
-      }
-    rescue => exc
-      error "failed to send message: #{exc.message}"
+      fail "not connected" unless @ws
+
+      case msg
+      when String
+        @ws.send(msg)
```

Comment:
TODO: consider having a separate `defer` thread or something to send these... these calls can block the actor, but they are also not thread-safe, because each `@ws.send` -> socket `write` may not be atomic with large messages?